File: lab6/java/task2/src/JewelryPackage/JewelryShop.java
package JewelryPackage;

import java.util.Scanner;

public class JewelryShop {
    private Jewelry jewelry;
    private ValuableJewelry valuableJewelry;
    private int count1;
    private int count2;
    private double extraPrice;
    private String name;

    public void init(String name, Jewelry j1, int c1, ValuableJewelry j2, int c2, double extraPrice) {
        this.jewelry = j1;
        this.valuableJewelry = j2;
        this.count1 = c1;
        this.count2 = c2;
        this.name = name;
        this.extraPrice = extraPrice;
    }

    public void read() {
        Scanner sc = new Scanner(System.in);
        System.out.print("Введите название магазина: ");
        this.name = sc.nextLine();

        System.out.print("Введите количество украшений 1: ");
        this.count1 = sc.nextInt();
        jewelry.read();

        System.out.print("Введите количество украшений 2: ");
        this.count2 = sc.nextInt();
        valuableJewelry.read();

        System.out.print("Введите стоимость доп. аксессуаров: ");
        this.extraPrice = sc.nextDouble();
    }

    public void display() {
        System.out.println("Название магазина: " + name);

        System.out.println("Обычное украшение: ");
        jewelry.display();
        System.out.println("Количество: " + count1);

        System.out.println("Ценное украшение: ");
        valuableJewelry.display();
        System.out.println("Количество: " + count2);

        System.out.println("Стоимость доп. аксессуаров: " + extraPrice);
    }

    public double getFullPrice() {
        return jewelry.getFullPricePerGram() * count1 +
                valuableJewelry.getFullPricePerGram() * count2 + +
                extraPrice;
    }

    public double getJewelryPrice() {
        return jewelry.getFullPricePerGram() +
                valuableJewelry.getFullPricePerGram();
    }
}

File: lab7/task2/cpp/JewelryShop.cpp
#include "Jewelry.cpp"
#include "ValuableJewelry.cpp"
#include <iostream>
#include <string> 

using namespace std;

// абстрактный основной класс
class JewelryShop {
  protected:
    int count1;
    int count2;
    double extraPrice;
    string* name;

  public:
    void init(string* name, int c1, int c2, double arg_extraPrice);

    void display();

    virtual double getFullPrice() = 0;

    virtual double getJewelryPrice() = 0;
};

// тип основного класса 1 - 2 базовых вспомогательных класса
class BaseShop: public JewelryShop {
  private:
    Jewelry j1;
    Jewelry j2;
  
  public:
    void init(string* arg_name, Jewelry arg_j1, int c1, Jewelry arg_j2, int c2, double arg_extraPrice);
    void display();
    double getFullPrice();
    double getJewelryPrice();
};

// тип основного класса 2 - 1 базовый и 1 производный классы
class MixedShop: public JewelryShop {
  private:
    Jewelry j1;
    ValuableJewelry j2;
  
  public:
    void init(string* arg_name, Jewelry arg_j1, int c1, ValuableJewelry arg_j2, int c2, double arg_extraPrice);
    void display();
    double getFullPrice();
    double getJewelryPrice();
};

// тип основного класса 3 - 2 производных класса
class ValuableShop: public JewelryShop {
  private:
    ValuableJewelry j1;
    ValuableJewelry j2;
  
  public:
    void init(string* arg_name, ValuableJewelry arg_j1, int c1, ValuableJewelry arg_j2, int c2, double arg_extraPrice);
    void display();
    double getFullPrice();
    double getJewelryPrice();
};


// определение методов


// методы абстрактного класса
void JewelryShop::init(string* arg_name, int c1, int c2, double arg_extraPrice) {
      count1 = c1;
      count2 = c2;

      name = arg_name;

      extraPrice = arg_extraPrice;
    };

void JewelryShop::display() {
  cout << "Название магазина: " << *(name) << endl;
  cout << "Стоимость доп. аксессуаров: " << extraPrice << endl;
  cout << "Количество украшений 1: " << count1 << endl;
  cout << "Количество украшений 2: " << count2 << endl;
};


// методы класса типа 1
void BaseShop::init(string* arg_name, Jewelry arg_j1, int c1, Jewelry arg_j2, int c2, double arg_extraPrice) {
  JewelryShop::init(arg_name, c1, c2, arg_extraPrice);
  j1 = arg_j1;
  j2 = arg_j2;
};

void BaseShop::display() {
  JewelryShop::display();
  cout << "Обычное украшение 1: " << endl;
  j1.display();
  cout << "Обычное украшение 2: " << endl;
  j2.display();
};

double BaseShop::getFullPrice() {
  return j1.getFullPricePerGramm() * count1 + j2.getFullPricePerGramm() * count2 + extraPrice;
};

double BaseShop::getJewelryPrice() {
  return j1.getFullPricePerGramm() + j2.getFullPricePerGramm();
};


// методы класса типа 2
void MixedShop::init(string* arg_name, Jewelry arg_j1, int c1, ValuableJewelry arg_j2, int c2, double arg_extraPrice) {
  JewelryShop::init(arg_name, c1, c2, arg_extraPrice);
  j1 = arg_j1;
  j2 = arg_j2;
};

void MixedShop::display() {
  JewelryShop::display();
  cout << "Обычное украшение 1: " << endl;
  j1.display();
  cout << "Ценное украшение 2: " << endl;
  j2.display();
};

double MixedShop::getFullPrice() {
  return j1.getFullPricePerGramm() * count1 + j2.getFullPricePerGramm() * count2 + extraPrice;
};

double MixedShop::getJewelryPrice() {
  return j1.getFullPricePerGramm() + j2.getFullPricePerGramm();
};


// методы класса типа 3
void ValuableShop::init(string* arg_name, ValuableJewelry arg_j1, int c1, ValuableJewelry arg_j2, int c2, double arg_extraPrice) {
  JewelryShop::init(arg_name, c1, c2, arg_extraPrice);
  j1 = arg_j1;
  j2 = arg_j2;
};

void ValuableShop::display() {
  JewelryShop::display();
  cout << "Ценное украшение 1: " << endl;
  j1.display();
  cout << "Ценное украшение 2: " << endl;
  j2.display();
};

double ValuableShop::getFullPrice() {
  return j1.getFullPricePerGramm() * count1 + j2.getFullPricePerGramm() * count2 + extraPrice;
};

double ValuableShop::getJewelryPrice() {
  return j1.getFullPricePerGramm() + j2.getFullPricePerGramm();
};
